### PR TITLE
TIS-574/journey pattern map

### PR DESCRIPTION
**Pull request by solita-sabinaf**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,7 +58,8 @@
     "react-router-dom": "6.26.0",
     "redux-sentry-middleware": "0.2.2",
     "uuid": "10.0.0",
-    "@fintraffic/fds-coreui-css": "0.1.3"
+    "@fintraffic/fds-coreui-css": "0.1.3",
+    "react-leaflet-cluster": "^2.1.0"
```
Comment (solita-sabinaf) [inline package.json:62]: !

File: src/api/index.tsx [hunk 2]
```diff
@@ -22,12 +22,24 @@ export type SearchForQuayResponse = {
...
+
+export type Location = {
+  longitude: number;
+  latitude: number;
 };
 
 export type StopPlace = {
   id: string;
   name: { value: string };
+  transportMode?: string;
```
Comment (solita-sabinaf) [inline src/api/index.tsx:41]: Use enum

File: src/components/JourneyPatternEditor/index.tsx [hunk 3]
```diff
@@ -11,13 +11,15 @@ import { useCallback, useState } from 'react';
...
 import './styles.scss';
+import { useConfig } from '../../config/ConfigContext';
 
 type Props = {
   journeyPattern: JourneyPattern;
   onSave: (journeyPattern: JourneyPattern) => void;
   onDelete?: () => void;
   spoilPristine: boolean;
   flexibleLineType?: FlexibleLineType;
+  transportMode?: string;
```
Comment (solita-sabinaf) [inline src/components/JourneyPatternEditor/index.tsx:22]: Enum

File: src/components/JourneyPatternEditor/index.tsx [hunk 4]
```diff
@@ -64,8 +75,34 @@ const JourneyPatternEditor = ({
...
+      let newServiceJourneys = serviceJourneys.map((serviceJourney) => ({
+        ...serviceJourney,
+        passingTimes: [...serviceJourney.passingTimes, {}],
+      }));
+
+      let newPointsInSequence = [
+        ...journeyPattern.pointsInSequence,
+        { quayRef },
+      ];
+
```
Comment (solita-sabinaf) [inline src/components/JourneyPatternEditor/index.tsx:96]: Add   window.dispatchEvent(new Event('resize'));

Comment (solita-sabinaf) [reply src/components/JourneyPatternEditor/index.tsx:96]: Move this method further to map component

File: src/components/StopPointsEditor/Generic/GenericStopPointsEditor.tsx [hunk 5]
```diff
@@ -40,16 +42,23 @@ export const GenericStopPointsEditor = ({
               onChange={(updatedStopPoint: StopPoint) =>
                 updateStopPoint(pointIndex, updatedStopPoint)
               }
-              onDelete={() => deleteStopPoint(pointIndex)}
+              onDelete={() => {
+                console.log('Delete button clicked', pointIndex);
```
Comment (solita-sabinaf) [inline src/components/StopPointsEditor/Generic/GenericStopPointsEditor.tsx:46]: !!!

File: src/components/StopPointsEditor/Generic/QuaysContext.tsx [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { createContext, useContext } from 'react';
```
Comment (solita-sabinaf) [inline src/components/StopPointsEditor/Generic/QuaysContext.tsx:1]: TO REMOVE

File: src/components/StopPointsEditor/index.tsx [hunk 7]
```diff
@@ -13,7 +13,9 @@ export type StopPointsEditorProps = {
   updateStopPoint: (index: number, stopPoint: StopPoint) => void;
   deleteStopPoint: (index: number) => void;
   addStopPoint: () => void;
+  addStopPointFromMap?: (quayRef: string) => void;
   onPointsInSequenceChange: (pointsInSequence: StopPoint[]) => void;
+  transportMode?: string;
```
Comment (solita-sabinaf) [inline src/components/StopPointsEditor/index.tsx:18]: enum

File: src/ext/JourneyPatternStopPointMap/JourneyPatternStopPointMap.tsx [hunk 8]
```diff
@@ -1,9 +1,177 @@
...
+      <MarkerClusterGroup chunkedLoading disableClusteringAtZoom={12}>
+        <Polyline positions={mapState.stopPointLocationSequence} />
+        {stopPlaces?.map((stopPlace: StopPlace) => {
+          const selectedQuayIds: string[] = stopPlace.quays
+            .filter(
+              (quay: Quay) =>
+                mapState.quayStopPointIndexRecord[quay.id] !== undefined,
+            )
+            .map((quay: Quay) => quay.id);
+          return mapState.showQuaysState[stopPlace.id] ? (
```
Comment (solita-sabinaf) [inline src/ext/JourneyPatternStopPointMap/JourneyPatternStopPointMap.tsx:126]: Move this to wrapper

File: src/ext/JourneyPatternStopPointMap/QuayMarker.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        stopPointIndex !== undefined,
+      )}
+      position={[
+        quay.centroid?.location.latitude,
+        quay.centroid?.location.longitude,
+      ]}
+    >
+      <Popup>
+        <section>
+          <h4 className={'popup-title'}>{stopPlace.name.value}</h4>
```
Comment (solita-sabinaf) [inline src/ext/JourneyPatternStopPointMap/QuayMarker.tsx:49]: Use header component

File: src/ext/JourneyPatternStopPointMap/QuayMarker.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+          {quay.name?.value && quay.name?.value !== stopPlace.name.value ? (
```
Comment (solita-sabinaf) [inline src/ext/JourneyPatternStopPointMap/QuayMarker.tsx:50]: Remove this until public code is available

File: src/ext/JourneyPatternStopPointMap/StopPlaceMarker.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+            )}
+          </section>
+        ) : (
+          <section>
+            <div>{formatMessage({ id: 'oneQuay' })}</div>
+            <div className={'popup-id'}>{stopPlace.quays[0].id}</div>
+          </section>
+        )}
+
+        <Button
```
Comment (solita-sabinaf) [inline src/ext/JourneyPatternStopPointMap/StopPlaceMarker.tsx:55]: 2 buttons

File: src/i18n/translations/nb.ts [hunk 12]
```diff
@@ -630,4 +630,14 @@ export const messages = {
...
+  // Journey patterns map, TODO:
+  showQuays: 'Show quays',
+  hideQuays: 'Hide quays',
+  oneQuay: '1 quay',
+  numberOfQuays: '{count} quays',
+  addToRoute: 'Add to route',
+  removeFromRoute: 'Remove from route',
+  quayTitle: 'Quay {name}',
+  showNonSelectedQuays: 'Show non-selected quays',
+  hideNonSelectedQuays: 'Hide non-selected quays',
```
Comment (solita-sabinaf) [inline src/i18n/translations/nb.ts:642]: !

File: src/reducers/stopPlaces.ts [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        stopPlaces.reduce((accumulator, currentStopPlace: StopPlace) => {
+          currentStopPlace.quays.forEach((quay) => {
+            accumulator[quay.id] = quay.centroid;
+          });
+          return accumulator;
+        }, {});
+
+      const quayStopPlaceIndex =
+        stopPlaces?.length > 0 &&
+        stopPlaces.reduce((accumulator, currentStopPlace: StopPlace) => {
```
Comment (solita-sabinaf) [inline src/reducers/stopPlaces.ts:31]: init accumulator

File: src/scenes/App/index.tsx [hunk 14]
```diff
@@ -32,7 +32,7 @@ L.Icon.Default.mergeOptions({
 const App = () => {
   const dispatch = useDispatch<any>();
   const auth = useAuth();
-  const { uttuApiUrl } = useConfig();
+  const { uttuApiUrl, sandboxFeatures } = useConfig();
```
Comment (solita-sabinaf) [inline src/scenes/App/index.tsx:35]: Remove

File: src/ext/JourneyPatternStopPointMap/QuayMarker.tsx [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+            }}
+            onClick={() => {
+              // If the quay got unselected and hideNonSelectedQuaysState mode was on, then switching to a stop place view:
+              if (stopPointIndex !== undefined && hideNonSelectedQuaysState) {
+                showQuaysCallback(false);
+                hideNonSelectedQuaysCallback(false);
+              }
+
+              // Callback to update the GenericStopPointsEditor form state:
+              if (stopPointIndex !== undefined) {
```
Comment (solita-sabinaf) [inline src/ext/JourneyPatternStopPointMap/QuayMarker.tsx:72]: This area gets refactored in the next PR about supporting multiple stop point indexes

File: src/i18n/translations/nb.ts [hunk 16]
```diff
@@ -630,4 +630,14 @@ export const messages = {
     'Din Nplan-installasjon trenger minst en dataleverandør',
   landingPageNotLoggedIn: 'Ikke logget inn',
   landingPageLoginButtonText: 'Logg inn',
+  // Journey patterns map, TODO:
```
Comment (testower) [inline src/i18n/translations/nb.ts:633]:
```suggestion
  showQuays: 'Vis quayer',
  hideQuays: 'Skjul quayer',
  oneQuay: '1 quay',
  numberOfQuays: '{count} quayer',
  addToRoute: 'Legg til',
  removeFromRoute: 'Fjern',
  quayTitle: 'Quay {name}',
  showNonSelectedQuays: 'Vis ikke valgte quayer,
  hideNonSelectedQuays: 'Skjul ikke valgte quayer',
```

File: src/i18n/translations/nb.ts (hunk 12 above)
Comment (testower) [reply src/i18n/translations/nb.ts:642]: General comment, I think maybe "route" here refers to the journey pattern is that correct?

Comment (solita-sabinaf) [reply src/i18n/translations/nb.ts:650]: That's true! Just journey pattern is a longer text for the button, but we can change to that if route is inappropriate

**Pull request merged**